### PR TITLE
Added reference_proxy implementation

### PR DESCRIPTION
**Pull request by JohanMabille**
Last piece missing from  #13

### CODE REVIEW

File: include/sparrow/array_data.hpp [hunk 1]
```diff
@@ -36,8 +50,251 @@ namespace sparrow
         std::vector<array_data> child_data;
     };
 
-    struct null_type
+    /**
+     * Reference type used in layout classes.
+     * This classes provides a similar behavior
+     * as std::optional, but on references in
```
Comment (Klaim) [inline include/sparrow/array_data.hpp:56]: Suggestion: `but as reference on data in an array_data`

File: test/test_array_data.cpp [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        }
+
+    private:
+
+        bool has_value(size_type index) const
+        {
+            return m_storage[index].has_value();
+        }
+
+        typename value_type::value_type& value(size_type index)
```
Comment (jjerphan) [inline test/test_array_data.cpp:55]: How about?
```suggestion
        typename inner_value_type& value(size_type index)
```

File: test/test_array_data.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        {
+            return m_storage[index].value();
+        }
+
+        const value_type::value_type& value(size_type index) const
```
Comment (jjerphan) [inline test/test_array_data.cpp:60]:
```suggestion
        const inner_value_type& value(size_type index) const
```

File: include/sparrow/array_data.hpp [hunk 4]
```diff
@@ -36,8 +50,251 @@ namespace sparrow
...
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(std::optional<U>&& rhs);
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(U&& value);
+
+        void reset();
+        void swap(self_type& rhs);
+
```
Comment (Klaim) [inline include/sparrow/array_data.hpp:92]: Consider adding a `// TODO: ...` for adding all the other usual arithmetic operators that might be allowed for `T`?

File: include/sparrow/array_data.hpp [hunk 5]
```diff
@@ -36,8 +50,251 @@ namespace sparrow
...
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(const std::optional<U>& rhs);
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(std::optional<U>&& rhs);
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(U&& value);
+
+        void reset();
```
Comment (Klaim) [inline include/sparrow/array_data.hpp:90]: It's not clear to me what this does exactly? Needs documentation, like `post-condition: has_value() == false` ?

File: test/test_array_data.cpp [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    public:
+
+        using value_type = std::optional<int>;
+        using inner_value_type = value_type::value_type;
+        using reference = reference_proxy<mock_layout>;
+        using size_type = std::size_t;
+        using storage_type = std::vector<std::optional<int>>;
+
+        mock_layout() = default;
+        mock_layout(std::initializer_list<value_type> l)
```
Comment (Klaim) [inline test/test_array_data.cpp:33]: `explicit` :D

File: test/test_array_data.cpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+{
+    class mock_layout
+    {
+    public:
+
+        using value_type = std::optional<int>;
+        using inner_value_type = value_type::value_type;
+        using reference = reference_proxy<mock_layout>;
+        using size_type = std::size_t;
+        using storage_type = std::vector<std::optional<int>>;
```
Comment (jjerphan) [inline test/test_array_data.cpp:30]:
```suggestion
        using storage_type = std::vector<value_type>;
```

**Review by Klaim — COMMENTED** [review]
Other than missing documentation  on functions and other details, LGTM

File: test/test_array_data.cpp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        {
+            int expected_value0 = stored_value(0);
+            int expected_value3 = stored_value(3);
+
+            auto ref0 = m_layout[0];
+            auto ref2 = m_layout[2];
+            auto ref3 = m_layout[3];
+
+            ref0.swap(ref2);
+        }
```
Comment (jjerphan) [inline test/test_array_data.cpp:171]: This test case looks incomplete. `ref3` is not needed.

File: test/test_array_data.cpp [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+            auto ref0 = m_layout[0];
+            CHECK_EQ(ref0.value(), stored_value(0));
+            static_assert(std::same_as<decltype(ref0.value()), int&>);
+
+            const auto cref0 = m_layout[0];
+            CHECK_EQ(cref0.value(), stored_value(0));
+            static_assert(std::same_as<decltype(cref0.value()), const int&>);
+        }
+
+        TEST_CASE_FIXTURE(ref_proxy_fixture, "assign")
```
Comment (jjerphan) [inline test/test_array_data.cpp:123]: We might test assigning `std::nullopt_t`.

**Review by jjerphan — COMMENTED** [review]
The implementation LGTM. I just have comments for the tests.

File: test/test_array_data.cpp (hunk 2 above)
Comment (JohanMabille) [reply test/test_array_data.cpp:55]: Ah indeed, I forgot to replace this. The typename is not even required with the change.

File: include/sparrow/array_data.hpp (hunk 4 above)
Comment (JohanMabille) [reply include/sparrow/array_data.hpp:92]: They don't exist for `std::optional`, that is why I haven't added them.

File: include/sparrow/array_data.hpp (hunk 5 above)
Comment (JohanMabille) [reply include/sparrow/array_data.hpp:90]: I was confused with `reset` from `std::optional`. I'll move this method to the private section.

File: test/test_array_data.cpp (hunk 6 above)
Comment (JohanMabille) [reply test/test_array_data.cpp:33]: not for initializer_list ;)

File: test/test_array_data.cpp (hunk 9 above)
Comment (JohanMabille) [reply test/test_array_data.cpp:123]: It is done line 130

File: include/sparrow/array_data.hpp (hunk 4 above)
Comment (Klaim) [reply include/sparrow/array_data.hpp:91]: Yes, I meant do we think we will need them in the context of arrow db manipulation?

File: include/sparrow/array_data.hpp [hunk 10]
```diff
@@ -36,8 +50,252 @@ namespace sparrow
         std::vector<array_data> child_data;
     };
 
-    struct null_type
+    /**
+     * Reference type used in layout classes.
+     * This classes provides a similar behavior
```
Comment (jjerphan) [inline include/sparrow/array_data.hpp:55]:
```suggestion
     * This class provides a similar behavior
```

**Review by jjerphan — APPROVED** [review]

File: include/sparrow/array_data.hpp (hunk 4 above)
Comment (JohanMabille) [reply include/sparrow/array_data.hpp:91]: For now I don't think so. most of the time, people simply perform the operations through the pair of value iterator / bitmask iterator, and perform operations even if the bitmask if false. If your data is not too sparse (which is the case most of the time), it is faster than testing the flag and branching for each value.

**Pull request merged**